/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.    
*/

syntax = "proto3";

option objc_class_prefix = "MAZE";

import "google/protobuf/empty.proto";

package maze;

// Service definition
service MazeWalker {
  // Send an event without data
  rpc GoWest (google.protobuf.Empty) returns (google.protobuf.Empty) {}
  rpc GoEast (google.protobuf.Empty) returns (google.protobuf.Empty) {}
  rpc GoNorth (google.protobuf.Empty) returns (google.protobuf.Empty) {}
  rpc GoSouth (google.protobuf.Empty) returns (google.protobuf.Empty) {}

  // Send an event with data
  rpc AdjustStepCount (AdjustStepCountRequest) returns (google.protobuf.Empty) {}

  // Invoke an event with reply data
  rpc StepCount (google.protobuf.Empty) returns (StepCountReply) {}

  // Subscribe for outgoing event with data
  rpc Subscribe_WrongWay (google.protobuf.Empty) returns (stream WrongWay) {}
  rpc Subscribe_GoalReached (google.protobuf.Empty) returns (stream StepCountReply) {} // Messages can be "reused"

  // Unsubscribe for outgoing event
  rpc Unsubscribe_WrongWay (google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

// Number of steps taken
message StepCountReply {
  int32 count = 1;
}

// Adjustment of the step counter
message AdjustStepCountRequest {
  int32 adjustment = 1;
}

// Outgoing event 
message WrongWay {
  string message = 1;
}
